File: co_robot/scripts/tcp_srv.py
#!/usr/bin/env python3
import rospy
from co_robot.srv import vision_robot
import socket
import time

SERVER_IP = "192.168.1.207"
SERVER_PORT = 9000
MAX_MESSAGE_SIZE = 100

sock = None
prev_x = None
prev_y = None
prev_z = None

def tcp_client_start():
    global sock
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        sock.connect((SERVER_IP, SERVER_PORT))
        print(f"Connected to TCP server at {SERVER_IP}:{SERVER_PORT}")
    except socket.error as e:
        print("연결 실패:", e)
        exit(1)

def tcp_client_send(msg):
    if len(msg) >= MAX_MESSAGE_SIZE:
        msg = msg[:MAX_MESSAGE_SIZE - 1]
    try:
        sock.sendall(msg.encode())
        print(f"Sent to server: {msg}")
    except socket.error as e:
        print("메시지 전송 실패:", e)

def tcp_client_receive():
    try:
        received = sock.recv(MAX_MESSAGE_SIZE).decode()
        if received:
            print(f"Received from server: {received}")
        return received
    except socket.error as e:
        print("응답 수신 실패:", e)
        return None

def robot_move(req):
    global prev_x, prev_y, prev_z

    val_1 = str(req.x)
    val_2 = str(req.y)
    val_3 = str(req.z)
    null = " "

    # (1) 좌표가 모두 0일 경우 무시
    if req.x == 0 and req.y == 0 and req.z == 0:
        return 100

    # (2) x 값이 허용 범위 벗어나면 무시
    if req.x > 590 or req.x < 298:
        rospy.loginfo("X value not sending the command")
        return 100

    # (3) 이전 값과 너무 비슷하면 무시
    if all(v is not None for v in [prev_x, prev_y, prev_z]):
        diff_x = abs(req.x - prev_x)
        diff_y = abs(req.y - prev_y)
        diff_z = abs(req.z - prev_z)
        if diff_x < 50 and diff_y < 50 and diff_z < 50:
            rospy.loginfo("이전값과 너무 유사하여 무시")
            return 100

    # (4) 명령 전송
    output = val_1 + "," + val_2 + "," + val_3 + "," + null
    print(f"명령 전송: {output}")

    received = tcp_client_receive()
    if received == "ok":
        tcp_client_send(output)

    time.sleep(1)

    global count_number
    count = str(count_number)

    received = tcp_client_receive()
    if received == "ok":
        tcp_client_send(count)

    count_number += 1

    # (5) 현재 값 갱신
    prev_x = req.x
    prev_y = req.y
    prev_z = req.z

    time.sleep(1)
    return 100

def main():
    global count_number
    count_number = 1

    tcp_client_start()
    time.sleep(1)

    rospy.init_node("tcp_srv")
    rospy.Service("service", vision_robot, robot_move)
    rospy.spin()

if __name__ == "__main__":
    main()

File: co_robot/src/tcp_srv.cpp
#include <iostream>
#include <typeinfo>
#include <cstring>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <stdio.h>

#include "ros/ros.h"
//#include "vision/vision_robot.h"

#define SERVER_IP "192.168.1.207"
#define SERVER_PORT 9000
#define MAX_MESSAGE_SIZE 100

int sock;
struct sockaddr_in server_addr;
char message[MAX_MESSAGE_SIZE];
int message_length;
int state = 1;
int count_number = 1;

void tcp_client_start() {

    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == -1) {
        perror("소켓 생성 실패");
        exit(1); 
    }

    // 서버 주소 설정
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = inet_addr(SERVER_IP);
    server_addr.sin_port = htons(SERVER_PORT);

    // 서버에 연결
    if (connect(sock, (struct sockaddr *)&server_addr, sizeof(server_addr)) == -1) {
        perror("연결 실패");
        exit(1);
    }
}

void tcp_client_send(std::string input){

    std::size_t length = input.size();
    if (length >= MAX_MESSAGE_SIZE) {
        length = MAX_MESSAGE_SIZE - 1;
    }
    std::strncpy(message, input.c_str(), length);

    message_length = strlen(message);
    // 개행 문자 제거
    if (message[message_length - 1] == '\n') {
        message[message_length - 1] = '\0';
        message_length--;
    }

    // 서버로 메시지 전송
    if (send(sock, message, message_length, 0) == -1) {
        perror("메시지 전송 실패");

    }
    
}

void tcp_client_read(std::string input){
    std::cout << "서버로부터 받은 응답: ";

    // 서버로부터 응답 받기
    memset(message, 0, sizeof(message));


    int received = recv(sock, message, sizeof(message), 0);


    if(received > 0){
        std::string str(message);
        std::cout << str << std::endl;


        if (str == "ok"){
            tcp_client_send(input);
        }

        if (str == "finish"){
            tcp_client_send(input);
        }
                
    }

}
    


bool robot_move(vision::vision_robot::Request &req, vision::vision_robot::Response &res){
    
    std::string val_1 = std::to_string(req.x);
    std::string val_2 = std::to_string(req.y);
    std::string val_3 = std::to_string(req.z);
    std::string null= " ";
    
    std::string output = val_1 + "," + val_2 + "," + val_3 + "," + null;
    std::cout<<output<<std::endl;

    tcp_client_read(output);
    sleep(1);

    std::string count = std::to_string(count_number);
    tcp_client_read(count);
    count_number += 1;

    
    sleep(1);
    res.state = 100;

    return true;
}

int main(int argc, char **argv){			
// Node Main Function

tcp_client_start();
sleep(1);

ros::init(argc, argv, "tcp_srv"); 	
ros::NodeHandle nh; 	

ros::ServiceServer server = nh.advertiseService("service", &robot_move);
// Initializes Node Name
// Node handle declaration for communication with ROS system
// Declares subscriber. Create subscriber 'ros_tutorial_sub' usythgbing the 'MsgTutorial'
// message file from the 'ros_tutorials_topic' package. The topic name is
// 'ros_tutorial_msg' and the size of the publisher queue is set to 100.

// tcp_client_read();
ros::spin();


// A function for calling a callback function, waiting for a message to be
// received, and executing a callback function when it is received
return 0;
}
